metrics/sari/sari.py

```diff
@@ -0,0 +1,285 @@
+    (keep3score, del3score, add3score) = SARIngram(s3grams, c3grams, r3gramslist, numref)
+    (keep4score, del4score, add4score) = SARIngram(s4grams, c4grams, r4gramslist, numref)
+    avgkeepscore = sum([keep1score, keep2score, keep3score, keep4score]) / 4
+    avgdelscore = sum([del1score, del2score, del3score, del4score]) / 4
+    avgaddscore = sum([add1score, add2score, add3score, add4score]) / 4
+    finalscore = (avgkeepscore + avgdelscore + avgaddscore) / 3
+    return finalscore
+
+
+def normalize(sentence, lowercase: bool = True, tokenizer: str = "13a", return_str: bool = True):
+
+    # Normalization is requried for the ASSET dataset (one of the primary
+    # datasets in sentence simplification) to allow using space
+    # to split the sentence. Even though Wiki-Auto and TURK datasets,
+    # do not require normalization, we do it for consistency.
+    # Code adapted from the EASSE library [1] written by the authors of the ASSET dataset.
+    # [1] https://github.com/feralvam/easse/blob/580bba7e1378fc8289c663f864e0487188fe8067/easse/utils/preprocessing.py#L7
+
+    if lowercase:
+        sentence = sentence.lower()
+
+    if tokenizer in ["13a", "intl"]:
+        normalized_sent = sacrebleu.TOKENIZERS[tokenizer]()(sentence)
+    elif tokenizer == "moses":
+        normalized_sent = sacremoses.MosesTokenizer().tokenize(sentence, return_str=True, escape=False)
+    elif tokenizer == "penn":
+        normalized_sent = sacremoses.MosesTokenizer().penn_tokenize(sentence, return_str=True)
+    else:
+        normalized_sent = sentence
+
+    if not return_str:
+        normalized_sent = normalized_sent.split()
+
+    return normalized_sent
+
+
+@datasets.utils.file_utils.add_start_docstrings(_DESCRIPTION, _KWARGS_DESCRIPTION)
+class Sari(datasets.Metric):
+    def _info(self):
+        return datasets.MetricInfo(
+            description=_DESCRIPTION,
+            citation=_CITATION,
+            inputs_description=_KWARGS_DESCRIPTION,
+            features=datasets.Features(
+                {
+                    "predictions": datasets.Value("string", id="sequence"),
+                    "references": datasets.Sequence(datasets.Value("string", id="sequence"), id="references"),
+                }
+            ),
+            codebase_urls=[
+                "https://github.com/cocoxu/simplification/blob/master/SARI.py",
+                "https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/utils/sari_hook.py",
+            ],
+            reference_urls=["https://www.aclweb.org/anthology/Q16-1029.pdf"],
+        )
+
+    def _compute(self, sources, predictions, references):
+        srcs = [normalize(sent) for sent in sources]
+        preds = [normalize(sent) for sent in predictions]
+        refs = [[normalize(sent) for sent in ref_sents] for ref_sents in references]
+
+        sari_score = []
+        for i in range(len(srcs)):
+            sari_score.append(SARIsent(srcs[i], preds[i], refs[i]))
```

Comment:
Maybe we could have something that doesn't store the normalized sentences to save memory ?
Also we can add a check that there are as many sources as predictions/references.

Maybe something like that ?

```suggestion
        if not (len(sources) == len(predictions) == len(references)):
            raise ValueError("Sources length must match predictions and references lengths.")

        sari_score = 0
        for src, pred, refs in zip(sources, predictions, references):
            sari_score += SARIsent(
                normalize(src), normalize(pred), [normalize(sent) for sent in refs]
            ))
        sari_score = sari_score / len(predictions)
```

Reply:
Thanks. I have modified the code based on the suggestions.